sycl/include/sycl/ext/oneapi/annotated_arg/annotated_arg.hpp

```diff
@@ -0,0 +1,97 @@
+//==----------- annotated_arg.hpp - SYCL annotated_arg extension -----------==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#include <cstddef>
+#include <type_traits>
+
+#include <sycl/detail/stl_type_traits.hpp>
+#include <sycl/exception.hpp>
+#include <sycl/ext/oneapi/annotated_arg/properties.hpp>
+#include <sycl/ext/oneapi/properties/properties.hpp>
+
+#ifdef __SYCL_DEVICE_ONLY__
+#define __SYCL_HOST_NOT_SUPPORTED(Op)
+#else
+#define __SYCL_HOST_NOT_SUPPORTED(Op)                                          \
+  throw sycl::exception(                                                       \
+      sycl::make_error_code(sycl::errc::feature_not_supported),                \
+      Op " is not supported on host device.");
+#endif
+
+namespace sycl {
+__SYCL_INLINE_VER_NAMESPACE(_V1) {
+namespace ext {
+namespace oneapi {
+namespace experimental {
+
+template <typename T, typename PropertyListT = detail::empty_properties_t>
+class annotated_arg {
+  // This should always fail when instantiating the unspecialized version.
+  static_assert(is_property_list<PropertyListT>::value,
+                "Property list is invalid.");
+};
+
+// Partial specialization to make PropertyListT visible as a parameter pack
+// of properties.
+template <typename T, typename... Props>
+class __attribute__((sycl_special_class)) annotated_arg<T, detail::properties_t<Props...>> {
+
+  using property_list_t = detail::properties_t<Props...>;
+
+  T obj;
+  #ifdef __SYCL_DEVICE_ONLY__
+    void __init(
+      [[__sycl_detail__::add_ir_attributes_kernel_parameter(
+          detail::PropertyMetaInfo<Props>::name...,
+          detail::PropertyMetaInfo<Props>::value...
+      )]]
+      T _obj) {
+        obj = _obj;
+    }
+  #endif
+
+public:
+  static_assert(std::is_trivially_copy_constructible<T>::value,
```

Comment:
The way kernel arguments work in SYCL requires `T` to be device copyable.  And `T` is device copyable when it's trivially copy constructible